File: backend/routes/regions.go
package routes

import (
	"starter/backend/handlers"
	"starter/backend/middleware"

	"github.com/gin-gonic/gin"
)

func setupRegionRoutes(rg *gin.RouterGroup) {
	// Stats
	rg.GET("/regions/stats", middleware.RequirePermission("regions.view"), handlers.GetRegionStats)

	// Provinces
	rg.GET("/regions/provinces", middleware.RequirePermission("regions.view"), handlers.GetProvinces)
	rg.GET("/regions/provinces/:id", middleware.RequirePermission("regions.view"), handlers.GetProvince)
	rg.POST("/regions/provinces", middleware.RequirePermission("regions.create"), handlers.CreateProvince)
	rg.PUT("/regions/provinces/:id", middleware.RequirePermission("regions.update"), handlers.UpdateProvince)

	// Regencies
	rg.GET("/regions/all-regencies", middleware.RequirePermission("regions.view"), handlers.GetAllRegencies)
	rg.GET("/regions/regencies/:province_id", middleware.RequirePermission("regions.view"), handlers.GetRegencies)
	rg.GET("/regions/regency/:id", middleware.RequirePermission("regions.view"), handlers.GetRegency)
	rg.POST("/regions/regencies", middleware.RequirePermission("regions.create"), handlers.CreateRegency)
	rg.PUT("/regions/regency/:id", middleware.RequirePermission("regions.update"), handlers.UpdateRegency)

	// Districts
	rg.GET("/regions/districts/:regency_id", middleware.RequirePermission("regions.view"), handlers.GetDistricts)
	rg.GET("/regions/district/:id", middleware.RequirePermission("regions.view"), handlers.GetDistrict)
	rg.POST("/regions/districts", middleware.RequirePermission("regions.create"), handlers.CreateDistrict)
	rg.PUT("/regions/district/:id", middleware.RequirePermission("regions.update"), handlers.UpdateDistrict)

	// Villages
	rg.GET("/regions/villages/:district_id", middleware.RequirePermission("regions.view"), handlers.GetVillages)
	rg.GET("/regions/village/:id", middleware.RequirePermission("regions.view"), handlers.GetVillage)
	rg.POST("/regions/villages", middleware.RequirePermission("regions.create"), handlers.CreateVillage)
	rg.PUT("/regions/village/:id", middleware.RequirePermission("regions.update"), handlers.UpdateVillage)
}

File: frontend/src/lib/api/regions.ts
import { api } from './client';

export interface Province {
  id: string;
  name: string;
}

export interface Regency {
  id: string;
  province_id: string;
  name: string;
  province?: Province;
  districts?: District[];
}

export interface District {
  id: string;
  regency_id: string;
  name: string;
  regency?: Regency;
  villages?: Village[];
}

export interface Village {
  id: string;
  district_id: string;
  name: string;
  district?: District;
}

export interface RegionStats {
  provinces: number;
  regencies: number;
  districts: number;
  villages: number;
}

export const regionsApi = {
  getStats: () => 
    api.get<{ data: RegionStats }>('/regions/stats'),
  
  // Provinces
  getProvinces: () => 
    api.get<{ data: Province[] }>('/regions/provinces'),
  
  getProvince: (id: string) => 
    api.get<{ data: Province & { regencies?: Regency[] } }>(`/regions/provinces/${id}`),
  
  createProvince: (data: { id: string; name: string }) =>
    api.post<{ data: Province; message: string }>('/regions/provinces', data),
  
  updateProvince: (id: string, data: { name: string }) =>
    api.put<{ data: Province; message: string }>(`/regions/provinces/${id}`, data),
  
  // Regencies
  getAllRegencies: () =>
    api.get<{ data: (Regency & { province?: Province })[] }>('/regions/all-regencies'),
  
  getRegencies: (provinceId: string) => 
    api.get<{ data: Regency[] }>(`/regions/regencies/${provinceId}`),
  
  getRegency: (id: string) => 
    api.get<{ data: Regency }>(`/regions/regency/${id}`),
  
  createRegency: (data: { id: string; province_id: string; name: string }) =>
    api.post<{ data: Regency; message: string }>('/regions/regencies', data),
  
  updateRegency: (id: string, data: { name: string }) =>
    api.put<{ data: Regency; message: string }>(`/regions/regency/${id}`, data),
  
  // Districts
  getDistricts: (regencyId: string) => 
    api.get<{ data: District[] }>(`/regions/districts/${regencyId}`),
  
  getDistrict: (id: string) => 
    api.get<{ data: District }>(`/regions/district/${id}`),
  
  createDistrict: (data: { id: string; regency_id: string; name: string }) =>
    api.post<{ data: District; message: string }>('/regions/districts', data),
  
  updateDistrict: (id: string, data: { name: string }) =>
    api.put<{ data: District; message: string }>(`/regions/district/${id}`, data),
  
  // Villages
  getVillages: (districtId: string) => 
    api.get<{ data: Village[] }>(`/regions/villages/${districtId}`),
  
  getVillage: (id: string) => 
    api.get<{ data: Village }>(`/regions/village/${id}`),
  
  createVillage: (data: { id: string; district_id: string; name: string }) =>
    api.post<{ data: Village; message: string }>('/regions/villages', data),
  
  updateVillage: (id: string, data: { name: string }) =>
    api.put<{ data: Village; message: string }>(`/regions/village/${id}`, data),
};
